File: Lab_7/j/Heap.java

class MaxHeap<T extends Comparable<T>> {
    // Instance Variables
    private T[] heap = null;
    private int length = 0;
    private int last = -1;

    // Private Static Methods
    private static int root() {
        return 0;
    }

    private static int up(int child) {
        return (child - 1) / 2;
    }

    private static int left(int root) {
        return 2 * root + 1;
    }

    private static int right(int root) {
        return 2 * root + 2;
    }

    // Private Instance Methods
    private boolean isLeaf(int root) {
        final int left = left(root);
        final int right = right(root);
        return left > last || (heap[left] == null && heap[right] == null);
    }

    private int posOfMax(int indices[]) {
        int max_i = indices[0];
        T max_T = heap[max_i];

        for (int i = 1; i < indices.length; i++) {
            final int index = indices[i];
            final T value = heap[index];

            if (value != null && max_T.compareTo(value) < 0) {
                max_i = index;
                max_T = value;
            }
        }

        return max_i;
    }

    private void swap(int a, int b) {
        T temp = heap[a];
        heap[a] = heap[b];
        heap[b] = temp;
    }

    @SuppressWarnings("unchecked")
    private void grow(int desired_index) {
        if (desired_index >= length) {
            length = right(length - 1) + 1;
            T temp[] = (T[])new Comparable[length];
            for (int i = 0; i <= last; i++) {
                temp[i] = heap[i];
            }
            heap = temp;
        }
    }

    private int heapify(int root) {
        if (isLeaf(root))
            return 0; // No change if leaf

        final int pos_of_max = this.posOfMax(new int[] { root, left(root), right(root) });

        if (pos_of_max != root) {
            this.swap(pos_of_max, root);
            return pos_of_max;
        }
        return 0; // no swap necessary
    }

    // Construtors
    @SuppressWarnings("unchecked")
    public MaxHeap() {
        this.last = -1;

        this.length = 3;

        this.heap = (T[])new Comparable[this.length];
    }

    @SuppressWarnings("unchecked")
    MaxHeap(T array[]) {
        this.last = array.length - 1;

        this.length = 0;
        while (this.length < this.last) {
            this.length = right(this.length);
        }
        this.length++;

        this.heap = (T[])new Comparable[this.length];

        for (int i = 0; i < array.length; i++) {
            this.heap[i] = array[i];
        }
        for (int i = array.length - 1; i >= 0; i--) {
            int swapped = heapify(i);
            while (swapped != 0)
                swapped = heapify(swapped);
        }
    }

    // Public Instance Methods
    T getMax() {
        return heap[0];
    }

    int getSize() {
        return length;
    }

    void insert(T key) {
        grow(last + 1);
        last++;
        heap[last] = key;

        int ancestor = up(last);
        while (heapify(ancestor)!=0) {
            // std::cout << " ";
            // tree_tools::tree_draw(heap, length, ancestor, " ");
            ancestor = up(ancestor);
        }
        ;
    }

    T remove() {
        if (last < 0) {
            throw new IndexOutOfBoundsException("cannot remove from empty heap.");
        }
        T out = heap[0];
        heap[0] = heap[last];
        heap[last] = null;
        last--;

        int swapped = heapify(0);
        while (swapped!=0) {
            swapped = heapify(swapped);
        }
        return out;
    }

    public String toString() {
        String out = "Heap: \n";
        out+=(heap[left(root())] != null || heap[right(root())] != null) ? "┬" : "─";
        out += tree_draw(root(), "");
        return out;
    }

    private String draw_branch(int head, String lpad, boolean end) {
        String out = lpad + (end ? "└" : "├");
        if (heap[head] == null) {
            out += "─\n";
        } else if (isLeaf(head)) {
            T value = heap[head];
            out += "─" + value + "\n";
        } else {
            out += "┬" + tree_draw(head, lpad + (end ? " " : "│"));
        }
        return out;
    }

    private String tree_draw(int head, String lpad) {
        String out = ""+heap[head]+"\n";
        if (!isLeaf(head)) {
            out += draw_branch(left(head), lpad, false);
            out += draw_branch(right(head), lpad, true);
        }
        return out;
    }

    public static void main(String[] args) {
        Integer arr[] = { 7, 17, 3, 5, 21, 27, 13 };
        MaxHeap<Integer> m = new MaxHeap<Integer>(arr);
        System.out.println(m);

        MaxHeap<Integer> n = new MaxHeap<>();
        for (Integer var : arr) {
            n.insert(var);
        }
        System.out.println(n);

        m.remove();
        System.out.println(m);
    }
};

// class Max_Heap{
// public:
// int size = 5;
// int *heap = new int[size];

// void maxheap_fromArray(int arr[], int n){
// this.heap = arr;
// this.size = n;
// int last = size;
// for(int i= last; i>=0 ; i-- ){
// this.heapify(i);
// }
// }

// void heapify(int index){
// }

// int get_max() {
// return heap[0];
// }

// int get_size() {
// return this.size;
// }

// void insert(int element);
// int remove();
// };

// int main() {
// int arr[] = {7, 17, 3, 5, 21, 27, 13};
// MaxHeap<int> m = MaxHeap<int>(arr);
// std::cout << m;

// MaxHeap<int> n = MaxHeap<int>();
// for (auto var : arr) {
// n.insert(var);
// }
// std::cout << n;

// m.remove();
// std::cout << m;
// }

File: Lab_7/c/Heap.cpp
#include <cstddef>
#include <iostream>

#include "array_tree_draw.h"

template <typename T>
class MaxHeap {
    // Instance Variables
    T** heap = nullptr;
    size_t length = 0;
    int last = -1;

    // Private Static Methods

    template <typename E>
    static size_t root();

    template <typename E>
    static size_t up(size_t child);

    template <typename E>
    static size_t left(size_t root);

    template <typename E>
    static size_t right(size_t root);

    static size_t root() {
        return 0;
    }

    static size_t up(size_t child) {
        return ((long)child - 1) / 2;
    }

    static size_t left(size_t root) {
        return 2 * root + 1;
    }

    static size_t right(size_t root) {
        return 2 * root + 2;
    }

    // Private Instance Methods

    template <typename E, size_t n>
    size_t posOfMax(size_t (&indices)[n]);

    template <typename E>
    void swap(size_t a, size_t b);

    template <typename E>
    void grow(size_t desired_index);

    template <typename E>
    void heapify(size_t root);

    template <size_t n>
    size_t posOfMax(size_t const (&indices)[n]) {
        size_t max_i = indices[0];
        T max_T = *heap[max_i];

        for (size_t i = 1; i < n; i++) {
            const auto index = indices[i];
            const auto* value = heap[index];

            if (value && max_T < *value) {
                max_i = index;
                max_T = *value;
            }
        }

        return max_i;
    }

    void swap(size_t a, size_t b) {
        auto* temp = heap[a];
        heap[a] = heap[b];
        heap[b] = temp;
    }

    void grow(size_t desired_index) {
        if (desired_index >= length) {
            length = right(length - 1) + 1;
            T** temp = new T* [length] { nullptr };
            for (size_t i = 0; i <= last; i++) {
                temp[i] = heap[i];
            }
            heap = temp;
        }
    }

    size_t heapify(size_t root) {
        const size_t indices[] = {root, left(root), right(root)};

        if (indices[1] > last || (!heap[indices[1]] && !heap[indices[2]]))
            return 0;  // No change if leaf

        const size_t pos_of_max = this->posOfMax(indices);

        if (pos_of_max != root) {
            this->swap(pos_of_max, root);
            return pos_of_max;
        }
        return 0;  // no swap necessary
    }

   public:
    // Construtors

    MaxHeap() {
        this->last = -1;

        this->length = 3;

        this->heap = new T* [this->length] { nullptr };
        // for (size_t i = 0; i < this->length; i++) {
        //     this->heap[i] = nullptr;
        // }
    }

    template <size_t n>
    MaxHeap(T (&array)[n]) {
        this->last = n - 1;

        this->length = 0;
        while (this->length < this->last) {
            this->length = right(this->length);
        }
        this->length++;

        this->heap = new T* [this->length] { nullptr };
        // for (size_t i = n; i < this->length; i++) {
        //     this->heap[i] = nullptr;
        // }
        for (size_t i = 0; i < n; i++) {
            this->heap[i] = new T{array[i]};
        }
        for (int i = (int)n - 1; i >= 0; i--) {
            size_t swapped = heapify(i);
            while (swapped)
                swapped = heapify(swapped);
        }
    }

    ~MaxHeap() {
        for (size_t i = 0; i < length; i++) {
            if (heap[i]) {
                delete heap[i];
                heap[i] = nullptr;
            }
        }
        delete[] heap;
        heap = nullptr;
    }

    // Public Instance Methods
    template <typename E>
    E getMax();
    template <typename E>
    size_t getSize();

    T getMax() {
        return *heap[0];
    }

    size_t getSize() {
        return length;
    }

    void insert(T& key) {
        grow(last + 1);
        last++;
        heap[last] = new T{key};

        size_t ancestor = up(last);
        while (heapify(ancestor)) {
            // std::cout << "    ";
            // tree_tools::tree_draw(heap, length, ancestor, "    ");
            ancestor = up(ancestor);
        };
    }

    T remove() {
        if (last < 0) {
            throw std::out_of_range("cannot remove from empty heap.");
        }
        T out = *heap[0];
        heap[0] = heap[last];
        heap[last] = nullptr;
        last--;

        size_t swapped = heapify(0);
        while (swapped) {
            swapped = heapify(swapped);
        }
        return out;
    }

    template <typename E>
    friend std::ostream& operator<<(std::ostream& o, MaxHeap<E>& m) {
        o << "Heap: " << std::endl;
        tree_tools::print(m.heap, m.length + 1);
        return o;
    }
};

// class Max_Heap{
// public:
// int size = 5;
// int *heap = new int[size];

// void maxheap_fromArray(int arr[], int n){
//     this->heap = arr;
//     this->size = n;
//     int last = size;
//     for(int i= last; i>=0 ; i-- ){
//         this->heapify(i);
//     }
// }

// void heapify(int index){
// }

// int get_max() {
//     return heap[0];
// }

// int get_size() {
//     return this->size;
// }

// void insert(int element);
// int remove();
// };

int main() {
    int arr[] = {7, 17, 3, 5, 21, 27, 13};
    MaxHeap<int> m = MaxHeap<int>(arr);
    std::cout << m << std::endl;

    MaxHeap<int> n = MaxHeap<int>();
    for (auto var : arr) {
        n.insert(var);
    }
    std::cout << n << std::endl;

    m.remove();
    std::cout << m << std::endl;
}
